R/uncount.R

```diff
@@ -23,15 +24,18 @@
 uncount <- function(data, weights, .remove = TRUE, .id = NULL) {
   weights_quo <- enquo(weights)
   w <- dplyr::pull(dplyr::mutate(data, `_weight` = !! weights_quo))
+  w <- vec_cast(w, integer(), x_arg = "weights")
 
-  if (!is.numeric(w)) {
-    stop("`weights` must evaluate to a numeric vector", call. = FALSE)
+  if (any(w < 0)) {
+    abort("all elements of `weights` must be >= 0")
   }
 
-  out <- data[rep(seq_nrow(data), w), , drop = FALSE]
+  out <- vec_slice(data, vec_rep_each(vec_seq_along(data), w))
 
-  if (.remove && quo_is_symbol(weights_quo)) {
-    out[[as_string(get_expr(weights_quo))]] <- NULL
+  weights_is_col <- quo_is_symbol(weights_quo) &&
```

Comment:
I don't really like the idea of making it easier to use variables outside of the data frame. I get that the current error message isn't very informative (and making it better would be a reasonable amount of work), but I'd prefer not to go in this direction.

Reply:
I am not sure I really understand. The behaviour regarding variables outside of the data frame is the same as before. I thought I need the flag `weights_is_col` for `dplyr::select()` but you pointed me in the direction of `any_of()` which is way simpler anyways.

Example of current behaviour
``` r
library(tidyr)

df <- tibble(x = c("a", "b"), n = c(1, 2))
m <- 2
uncount(df, m)
#> # A tibble: 4 x 2
#>   x         n
#>   <chr> <dbl>
#> 1 a         1
#> 2 a         1
#> 3 b         2
#> 4 b         2
```

<sup>Created on 2020-12-07 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
Yeah, sorry, that was a red herring.